ICE `adjust mode unimplemented for ConstBlock`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_hir_typeck/src/pat.rs:689:21: FIXME(deref_patterns): adjust mode unimplemented for ConstBlock(ConstBlock { hir_id: HirId(DefId(0:3 ~ a[df51]::main).25), def_id: DefId(0:4 ~ a[df51]::main::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:3 ~ a[df51]::main).26) } })', 'error: internal compiler error: compiler/rustc_hir_typeck/src/pat.rs:689:21: FIXME(deref_patterns): adjust mode unimplemented for ConstBlock(ConstBlock { hir_id: HirId(DefId(0:3 ~ a[df51]::main).25), def_id: DefId(0:4 ~ a[df51]::main::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:3 ~ a[df51]::main).26) } })'
File: /tmp/im/2/a.rs
-->

auto-reduced (treereduce-rust):
````rust
#![feature(deref_patterns)]
fn main() {
    let vec![const { vec![] }]: Vec<usize> = vec![];
}

````

original:
````rust
#![feature(deref_patterns)]
fn main() {
    let vec![const { vec![] }]: Vec<usize> = vec![];

    {
        vec![]
    }
}

````

Version information
````
rustc 1.93.0-nightly (a8664a153 2025-10-26)
binary: rustc
commit-hash: a8664a1534913ccff491937ec2dc7ec5d973c2bd
commit-date: 2025-10-26
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.3

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs#L683-L695

Command:
`/home/matthias/.rustup/toolchains/alt-master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0532]: expected a pattern, found a function call
 --> /tmp/icemaker_global_tempdir.mXNDgDYHM4YL/rustc_testrunner_tmpdir_reporting.6gH2c4TMrPNV/mvce.rs:3:9
  |
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
  |         ^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
  |
  = note: function calls are not allowed in patterns: <https://doc.rust-lang.org/book/ch19-00-patterns.html>
  = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0658]: usage of qualified paths in this context is experimental
 --> /tmp/icemaker_global_tempdir.mXNDgDYHM4YL/rustc_testrunner_tmpdir_reporting.6gH2c4TMrPNV/mvce.rs:3:9
  |
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
  |         ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #86935 <https://github.com/rust-lang/rust/issues/86935> for more information
  = help: add `#![feature(more_qualified_paths)]` to the crate attributes to enable
  = note: this compiler was built on 2025-10-26; consider upgrading it if it is out of date
  = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: the feature `deref_patterns` is incomplete and may not be safe to use and/or cause compiler crashes
 --> /tmp/icemaker_global_tempdir.mXNDgDYHM4YL/rustc_testrunner_tmpdir_reporting.6gH2c4TMrPNV/mvce.rs:1:12
  |
1 | #![feature(deref_patterns)]
  |            ^^^^^^^^^^^^^^
  |
  = note: see issue #87121 <https://github.com/rust-lang/rust/issues/87121> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0164]: expected tuple struct or tuple variant, found associated function `<[_]>::into_vec`
 --> /tmp/icemaker_global_tempdir.mXNDgDYHM4YL/rustc_testrunner_tmpdir_reporting.6gH2c4TMrPNV/mvce.rs:3:9
  |
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
  |         ^^^^^^^^^^^^^^^^^^^^^^ `fn` calls are not allowed in patterns
  |
  = help: for more information, visit https://doc.rust-lang.org/book/ch19-00-patterns.html
  = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

error: internal compiler error: compiler/rustc_hir_typeck/src/pat.rs:689:21: FIXME(deref_patterns): adjust mode unimplemented for ConstBlock(ConstBlock { hir_id: HirId(DefId(0:3 ~ mvce[706e]::main).25), def_id: DefId(0:4 ~ mvce[706e]::main::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:3 ~ mvce[706e]::main).26) } })
 --> /tmp/icemaker_global_tempdir.mXNDgDYHM4YL/rustc_testrunner_tmpdir_reporting.6gH2c4TMrPNV/mvce.rs:3:14
  |
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
  |              ^^^^^^^^^^^^^^^^


thread 'rustc' (2804419) panicked at compiler/rustc_hir_typeck/src/pat.rs:689:21:
Box<dyn Any>
stack backtrace:
   0:     0x7f1cc72b0973 - std::backtrace_rs::backtrace::libunwind::trace::heae8212c72c002b5
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/../../backtrace/src/backtrace/libunwind.rs:117:9
   1:     0x7f1cc72b0973 - std::backtrace_rs::backtrace::trace_unsynchronized::h476583f2b33a8ed0
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/../../backtrace/src/backtrace/mod.rs:66:14
   2:     0x7f1cc72b0973 - std::sys::backtrace::_print_fmt::hc836bf0fe40bb1db
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/backtrace.rs:66:9
   3:     0x7f1cc72b0973 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hf84e64cdd27a2a23
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/backtrace.rs:39:26
   4:     0x7f1cc2635702 - core::fmt::rt::Argument::fmt::h8860ce552ce5bc30
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/core/src/fmt/rt.rs:173:76
   5:     0x7f1cc2635702 - core::fmt::write::h11a59be2c6656721
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/core/src/fmt/mod.rs:1469:25
   6:     0x7f1cc723c609 - std::io::default_write_fmt::hada6ba6c4ae80335
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/io/mod.rs:639:11
   7:     0x7f1cc723c609 - std::io::Write::write_fmt::h63b115eaf01ccee3
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/io/mod.rs:1954:13
   8:     0x7f1cc726189b - std::sys::backtrace::BacktraceLock::print::h4436f6d63c04792d
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/backtrace.rs:42:9
   9:     0x7f1cc726189b - std::panicking::default_hook::{{closure}}::hf85fe377ebbce52f
  10:     0x7f1cc7261435 - std::panicking::default_hook::h341448ad7b6c7250
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:328:9
  11:     0x7f1cc3a48f12 - <alloc[51e647838ce4ca05]::boxed::Box<dyn for<'a, 'b> core[2c4661bac6f6ac6]::ops::function::Fn<(&'a std[64cbbcbf1b940fdd]::panic::PanicHookInfo<'b>,), Output = ()> + core[2c4661bac6f6ac6]::marker::Sync + core[2c4661bac6f6ac6]::marker::Send> as core[2c4661bac6f6ac6]::ops::function::Fn<(&std[64cbbcbf1b940fdd]::panic::PanicHookInfo,)>>::call
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/alloc/src/boxed.rs:2019:9
  12:     0x7f1cc3a48f12 - rustc_driver_impl[c073ed665df9c1b8]::install_ice_hook::{closure#1}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_driver_impl/src/lib.rs:1430:17
  13:     0x7f1cc3a48f12 - <alloc[51e647838ce4ca05]::boxed::Box<rustc_driver_impl[c073ed665df9c1b8]::install_ice_hook::{closure#1}> as core[2c4661bac6f6ac6]::ops::function::Fn<(&dyn for<'a, 'b> core[2c4661bac6f6ac6]::ops::function::Fn<(&'a std[64cbbcbf1b940fdd]::panic::PanicHookInfo<'b>,), Output = ()> + core[2c4661bac6f6ac6]::marker::Sync + core[2c4661bac6f6ac6]::marker::Send, &std[64cbbcbf1b940fdd]::panic::PanicHookInfo)>>::call
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/alloc/src/boxed.rs:2019:9
  14:     0x7f1cc3a48f12 - std[64cbbcbf1b940fdd]::panicking::update_hook::<alloc[51e647838ce4ca05]::boxed::Box<rustc_driver_impl[c073ed665df9c1b8]::install_ice_hook::{closure#1}>>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:243:47
  15:     0x7f1cc7261eb4 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h656d4e94e6d8d203
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/alloc/src/boxed.rs:2019:9
  16:     0x7f1cc7261eb4 - std::panicking::panic_with_hook::h9a8f9bc8d4484163
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:842:13
  17:     0x7f1cc3b18c93 - std[64cbbcbf1b940fdd]::panicking::begin_panic::<rustc_errors[6510cf1d96c99ae7]::ExplicitBug>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:771:9
  18:     0x7f1cc3af4886 - std[64cbbcbf1b940fdd]::sys::backtrace::__rust_end_short_backtrace::<std[64cbbcbf1b940fdd]::panicking::begin_panic<rustc_errors[6510cf1d96c99ae7]::ExplicitBug>::{closure#0}, !>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/backtrace.rs:174:18
  19:     0x7f1cc3ae7089 - std[64cbbcbf1b940fdd]::panicking::begin_panic::<rustc_errors[6510cf1d96c99ae7]::ExplicitBug>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:770:5
  20:     0x7f1cc3b68e11 - std[64cbbcbf1b940fdd]::panic::panic_any::<rustc_errors[6510cf1d96c99ae7]::ExplicitBug>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panic.rs:260:5
  21:     0x7f1cc3b68e11 - <rustc_errors[6510cf1d96c99ae7]::diagnostic::BugAbort as rustc_errors[6510cf1d96c99ae7]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_errors/src/diagnostic.rs:63:9
  22:     0x7f1cc4feb4bc - <rustc_errors[6510cf1d96c99ae7]::diagnostic::Diag<rustc_errors[6510cf1d96c99ae7]::diagnostic::BugAbort>>::emit
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_errors/src/diagnostic.rs:1405:9
  23:     0x7f1cc4feb4bc - <rustc_errors[6510cf1d96c99ae7]::DiagCtxtHandle>::span_bug::<rustc_span[2c1246c92d8898ab]::span_encoding::Span, alloc[51e647838ce4ca05]::string::String>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_errors/src/lib.rs:1261:48
  24:     0x7f1cc50876c5 - rustc_middle[3b90aebf889658c7]::util::bug::opt_span_bug_fmt::<rustc_span[2c1246c92d8898ab]::span_encoding::Span>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/util/bug.rs:38:54
  25:     0x7f1cc508799a - rustc_middle[3b90aebf889658c7]::ty::context::tls::with_opt::<rustc_middle[3b90aebf889658c7]::util::bug::opt_span_bug_fmt<rustc_span[2c1246c92d8898ab]::span_encoding::Span>::{closure#0}, !>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/ty/context/tls.rs:136:23
  26:     0x7f1cc5028c16 - rustc_middle[3b90aebf889658c7]::ty::context::tls::with_context_opt::<rustc_middle[3b90aebf889658c7]::ty::context::tls::with_opt<rustc_middle[3b90aebf889658c7]::util::bug::opt_span_bug_fmt<rustc_span[2c1246c92d8898ab]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/ty/context/tls.rs:79:18
  27:     0x7f1cc5025eea - rustc_middle[3b90aebf889658c7]::ty::context::tls::with_opt::<rustc_middle[3b90aebf889658c7]::util::bug::opt_span_bug_fmt<rustc_span[2c1246c92d8898ab]::span_encoding::Span>::{closure#0}, !>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/ty/context/tls.rs:134:5
  28:     0x7f1cc5025eea - rustc_middle[3b90aebf889658c7]::util::bug::opt_span_bug_fmt::<rustc_span[2c1246c92d8898ab]::span_encoding::Span>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/util/bug.rs:33:5
  29:     0x7f1cc5025eea - rustc_middle[3b90aebf889658c7]::util::bug::span_bug_fmt::<rustc_span[2c1246c92d8898ab]::span_encoding::Span>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/util/bug.rs:24:5
  30:     0x7f1cc45f2fa7 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::calc_adjust_mode
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:689:21
  31:     0x7f1cc45f2fa7 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:402:32
  32:     0x7f1cc45e0e94 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_slice
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:2912:18
  33:     0x7f1cc45e0e94 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_inner
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:645:22
  34:     0x7f1cc45f293b - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:403:23
  35:     0x7f1cc45dde92 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_inner
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:601:30
  36:     0x7f1cc45f293b - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:403:23
  37:     0x7f1cc45dde92 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_inner
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:601:30
  38:     0x7f1cc45de5d0 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_inner
  39:     0x7f1cc45f293b - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:403:23
  40:     0x7f1cc454f223 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_pat_top
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/pat.rs:382:14
  41:     0x7f1cc454f223 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_decl
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:889:14
  42:     0x7f1cc4554d40 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_decl_local
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:910:23
  43:     0x7f1cc4554d40 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_stmt
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:934:22
  44:     0x7f1cc4554d40 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_block::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1015:22
  45:     0x7f1cc4554d40 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::with_breakable_ctxt::<<rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_block::{closure#0}, ()>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1527:22
  46:     0x7f1cc4554d40 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_block
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1013:31
  47:     0x7f1cc446e927 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_kind
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:574:46
  48:     0x7f1cc44a8bc8 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:282:23
  49:     0x7f1cc44a8bc8 - stacker[c1c5f7b2586f5440]::maybe_grow::<rustc_middle[3b90aebf889658c7]::ty::Ty, <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /rust/deps/stacker-0.1.21/src/lib.rs:57:9
  50:     0x7f1cc44a8bc8 - rustc_data_structures[623314d6c4688740]::stack::ensure_sufficient_stack::<rustc_middle[3b90aebf889658c7]::ty::Ty, <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_data_structures/src/stack.rs:21:5
  51:     0x7f1cc44a8bc8 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:277:18
  52:     0x7f1cc44a6333 - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_expectation
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:222:14
  53:     0x7f1cc44a5b3a - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_expr_with_hint
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:177:14
  54:     0x7f1cc44a5b3a - <rustc_hir_typeck[1949be542646bfbe]::fn_ctxt::FnCtxt>::check_return_or_body_tail
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/expr.rs:1133:35
  55:     0x7f1cc4641aae - rustc_hir_typeck[1949be542646bfbe]::check::check_fn
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/check.rs:130:9
  56:     0x7f1cc4437ec5 - rustc_hir_typeck[1949be542646bfbe]::typeck_with_inspect::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/lib.rs:182:9
  57:     0x7f1cc4459ba6 - rustc_hir_typeck[1949be542646bfbe]::typeck_with_inspect
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_hir_typeck/src/lib.rs:107:1
  58:     0x7f1cc5f08ebe - rustc_query_impl[629cfdbba787a83d]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_query_impl/src/plumbing.rs:294:9
  59:     0x7f1cc5f08ebe - rustc_query_impl[629cfdbba787a83d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[629cfdbba787a83d]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_query_impl/src/plumbing.rs:578:18
  60:     0x7f1cc63b692f - rustc_query_impl[629cfdbba787a83d]::query_impl::typeck::dynamic_query::{closure#2}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_query_impl/src/plumbing.rs:655:25
  61:     0x7f1cc63b692f - <rustc_query_impl[629cfdbba787a83d]::query_impl::typeck::dynamic_query::{closure#2} as core[2c4661bac6f6ac6]::ops::function::FnOnce<(rustc_middle[3b90aebf889658c7]::ty::context::TyCtxt, rustc_span[2c1246c92d8898ab]::def_id::LocalDefId)>>::call_once
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/core/src/ops/function.rs:250:5
  62:     0x7f1cc60296e6 - <rustc_query_impl[629cfdbba787a83d]::DynamicConfig<rustc_data_structures[623314d6c4688740]::vec_cache::VecCache<rustc_hir_id[9d51473d61962b3]::OwnerId, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[585d893f3b83678a]::dep_graph::graph::DepNodeIndex>, false, false, false> as rustc_query_system[585d893f3b83678a]::query::config::QueryConfig<rustc_query_impl[629cfdbba787a83d]::plumbing::QueryCtxt>>::compute
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_query_impl/src/lib.rs:122:9
  63:     0x7f1cc60296e6 - rustc_query_system[585d893f3b83678a]::query::plumbing::execute_job_non_incr::<rustc_query_impl[629cfdbba787a83d]::DynamicConfig<rustc_data_structures[623314d6c4688740]::vec_cache::VecCache<rustc_hir_id[9d51473d61962b3]::OwnerId, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[585d893f3b83678a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[629cfdbba787a83d]::plumbing::QueryCtxt>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_query_system/src/query/plumbing.rs:493:72
  64:     0x7f1cc60296e6 - rustc_middle[3b90aebf889658c7]::ty::context::tls::enter_context::<rustc_query_system[585d893f3b83678a]::query::plumbing::execute_job_non_incr<rustc_query_impl[629cfdbba787a83d]::DynamicConfig<rustc_data_structures[623314d6c4688740]::vec_cache::VecCache<rustc_hir_id[9d51473d61962b3]::OwnerId, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[585d893f3b83678a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[629cfdbba787a83d]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>>::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/compiler/rustc_middle/src/ty/context/tls.rs:60:9
  65:     0x7f1cc60296e6 - <std[64cbbcbf1b940fdd]::thread::local::LocalKey<core[2c4661bac6f6ac6]::cell::Cell<*const ()>>>::try_with::<rustc_middle[3b90aebf889658c7]::ty::context::tls::enter_context<rustc_query_system[585d893f3b83678a]::query::plumbing::execute_job_non_incr<rustc_query_impl[629cfdbba787a83d]::DynamicConfig<rustc_data_structures[623314d6c4688740]::vec_cache::VecCache<rustc_hir_id[9d51473d61962b3]::OwnerId, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[585d893f3b83678a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[629cfdbba787a83d]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>>::{closure#0}, rustc_middle[3b90aebf889658c7]::query::erase::Erased<[u8; 8usize]>>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/thread/local.rs:508:12
  
< snip > 

 154:     0x7f1cc3a33263 - std[64cbbcbf1b940fdd]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/backtrace.rs:158:18
 155:     0x7f1cc3a50f64 - <std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_::<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/thread/mod.rs:562:17
 156:     0x7f1cc3a50f64 - <core[2c4661bac6f6ac6]::panic::unwind_safe::AssertUnwindSafe<<std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}> as core[2c4661bac6f6ac6]::ops::function::FnOnce<()>>::call_once
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/core/src/panic/unwind_safe.rs:274:9
 157:     0x7f1cc3a50f64 - std[64cbbcbf1b940fdd]::panicking::catch_unwind::do_call::<core[2c4661bac6f6ac6]::panic::unwind_safe::AssertUnwindSafe<<std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:590:40
 158:     0x7f1cc3a50f64 - std[64cbbcbf1b940fdd]::panicking::catch_unwind::<(), core[2c4661bac6f6ac6]::panic::unwind_safe::AssertUnwindSafe<<std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panicking.rs:553:19
 159:     0x7f1cc3a50f64 - std[64cbbcbf1b940fdd]::panic::catch_unwind::<core[2c4661bac6f6ac6]::panic::unwind_safe::AssertUnwindSafe<<std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/panic.rs:359:14
 160:     0x7f1cc3a50f64 - <std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_::<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/thread/mod.rs:560:30
 161:     0x7f1cc3a50f64 - <<std[64cbbcbf1b940fdd]::thread::Builder>::spawn_unchecked_<rustc_interface[35bb7f6c97091001]::util::run_in_thread_with_globals<rustc_interface[35bb7f6c97091001]::util::run_in_thread_pool_with_globals<rustc_interface[35bb7f6c97091001]::interface::run_compiler<(), rustc_driver_impl[c073ed665df9c1b8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[2c4661bac6f6ac6]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
                               at /rustc-dev/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/core/src/ops/function.rs:250:5
 162:     0x7f1cc7251b85 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb679a76bc5040637
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/alloc/src/boxed.rs:2005:9
 163:     0x7f1cc7251b85 - std::sys::thread::unix::Thread::new::thread_start::h84b2d2778567a179
                               at /rustc/a8664a1534913ccff491937ec2dc7ec5d973c2bd/library/std/src/sys/thread/unix.rs:126:17
 164:     0x7f1cc0a969cb - <unknown>
 165:     0x7f1cc0b1aa0c - <unknown>
 166:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.93.0-nightly (a8664a153 2025-10-26) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 4 previous errors; 1 warning emitted

Some errors have detailed explanations: E0164, E0532, E0658.
For more information about an error, try `rustc --explain E0164`.

```

</p>
</details>

<!--
query stack:
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
3 |     let vec![const { vec![] }]: Vec<usize> = vec![];
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on this crate

-->
@rustbot label +F-deref_patterns

--- a/compiler/rustc_ast_lowering/messages.ftl
+++ b/compiler/rustc_ast_lowering/messages.ftl
@@ -6,6 +6,7 @@ ast_lowering_abi_specified_multiple_times =
 ast_lowering_arbitrary_expression_in_pattern =
     arbitrary expressions aren't allowed in patterns
     .pattern_from_macro_note = the `expr` fragment specifier forces the metavariable's content to be an expression
+    .const_block_in_pattern_help = use a named `const`-item or an `if`-guard (`x if x == const {"{ ... }"}`) instead
 
 ast_lowering_argument = argument
 

--- a/compiler/rustc_ast_lowering/src/errors.rs
+++ b/compiler/rustc_ast_lowering/src/errors.rs
@@ -357,6 +357,8 @@ pub(crate) struct ArbitraryExpressionInPattern {
     pub span: Span,
     #[note(ast_lowering_pattern_from_macro_note)]
     pub pattern_from_macro_note: bool,
+    #[help(ast_lowering_const_block_in_pattern_help)]
+    pub const_block_in_pattern_help: bool,
 }
 
 #[derive(Diagnostic)]

--- a/compiler/rustc_ast_lowering/src/pat.rs
+++ b/compiler/rustc_ast_lowering/src/pat.rs
@@ -399,7 +399,6 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             ExprKind::Lit(lit) => {
                 hir::PatExprKind::Lit { lit: self.lower_lit(lit, span), negated: false }
             }
-            ExprKind::ConstBlock(c) => hir::PatExprKind::ConstBlock(self.lower_const_block(c)),
             ExprKind::IncludedBytes(byte_sym) => hir::PatExprKind::Lit {
                 lit: respan(span, LitKind::ByteStr(*byte_sym, StrStyle::Cooked)),
                 negated: false,
@@ -419,10 +418,12 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 hir::PatExprKind::Lit { lit: self.lower_lit(lit, span), negated: true }
             }
             _ => {
+                let is_const_block = matches!(expr.kind, ExprKind::ConstBlock(_));
                 let pattern_from_macro = expr.is_approximately_pattern();
                 let guar = self.dcx().emit_err(ArbitraryExpressionInPattern {
                     span,
                     pattern_from_macro_note: pattern_from_macro,
+                    const_block_in_pattern_help: is_const_block,
                 });
                 err(guar)
             }

--- a/compiler/rustc_hir/src/hir.rs
+++ b/compiler/rustc_hir/src/hir.rs
@@ -1921,7 +1921,6 @@ pub enum PatExprKind<'hir> {
         // once instead of matching on unop neg expressions everywhere.
         negated: bool,
     },
-    ConstBlock(ConstBlock),
     /// A path pattern for a unit struct/variant or a (maybe-associated) constant.
     Path(QPath<'hir>),
 }

--- a/compiler/rustc_hir/src/intravisit.rs
+++ b/compiler/rustc_hir/src/intravisit.rs
@@ -789,7 +789,6 @@ pub fn walk_pat_expr<'v, V: Visitor<'v>>(visitor: &mut V, expr: &'v PatExpr<'v>)
     try_visit!(visitor.visit_id(*hir_id));
     match kind {
         PatExprKind::Lit { lit, negated } => visitor.visit_lit(*hir_id, *lit, *negated),
-        PatExprKind::ConstBlock(c) => visitor.visit_inline_const(c),
         PatExprKind::Path(qpath) => visitor.visit_qpath(qpath, *hir_id, *span),
     }
 }

--- a/compiler/rustc_hir_pretty/src/lib.rs
+++ b/compiler/rustc_hir_pretty/src/lib.rs
@@ -1876,7 +1876,6 @@ impl<'a> State<'a> {
                 }
                 self.print_literal(lit);
             }
-            hir::PatExprKind::ConstBlock(c) => self.print_inline_const(c),
             hir::PatExprKind::Path(qpath) => self.print_qpath(qpath, true),
         }
     }

--- a/compiler/rustc_hir_typeck/src/pat.rs
+++ b/compiler/rustc_hir_typeck/src/pat.rs
@@ -925,9 +925,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
                 ty
             }
-            rustc_hir::PatExprKind::ConstBlock(c) => {
-                self.check_expr_const_block(c, Expectation::NoExpectation)
-            }
             rustc_hir::PatExprKind::Path(qpath) => {
                 let (res, opt_ty, segments) =
                     self.resolve_ty_and_res_fully_qualified_call(qpath, lt.hir_id, lt.span);

--- a/compiler/rustc_mir_build/src/thir/pattern/mod.rs
+++ b/compiler/rustc_mir_build/src/thir/pattern/mod.rs
@@ -13,14 +13,13 @@ use rustc_hir::def::{CtorOf, DefKind, Res};
 use rustc_hir::pat_util::EnumerateAndAdjustIterator;
 use rustc_hir::{self as hir, ByRef, LangItem, Mutability, Pinnedness, RangeEnd};
 use rustc_index::Idx;
-use rustc_infer::infer::TyCtxtInferExt;
 use rustc_middle::mir::interpret::LitToConstInput;
 use rustc_middle::thir::{
     Ascription, FieldPat, LocalVarId, Pat, PatKind, PatRange, PatRangeBoundary,
 };
 use rustc_middle::ty::adjustment::{PatAdjust, PatAdjustment};
 use rustc_middle::ty::layout::IntegerExt;
-use rustc_middle::ty::{self, CanonicalUserTypeAnnotation, Ty, TyCtxt, TypingMode};
+use rustc_middle::ty::{self, CanonicalUserTypeAnnotation, Ty, TyCtxt};
 use rustc_middle::{bug, span_bug};
 use rustc_span::def_id::DefId;
 use rustc_span::{ErrorGuaranteed, Span};
@@ -621,54 +620,8 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
         pattern
     }
 
-    /// Lowers an inline const block (e.g. `const { 1 + 1 }`) to a pattern.
-    fn lower_inline_const(
-        &mut self,
-        block: &'tcx hir::ConstBlock,
-        id: hir::HirId,
-        span: Span,
-    ) -> PatKind<'tcx> {
-        let tcx = self.tcx;
-        let def_id = block.def_id;
-        let ty = tcx.typeck(def_id).node_type(block.hir_id);
-
-        let typeck_root_def_id = tcx.typeck_root_def_id(def_id.to_def_id());
-        let parent_args = ty::GenericArgs::identity_for_item(tcx, typeck_root_def_id);
-        let args = ty::InlineConstArgs::new(tcx, ty::InlineConstArgsParts { parent_args, ty }).args;
-
-        let ct = ty::UnevaluatedConst { def: def_id.to_def_id(), args };
-        let c = ty::Const::new_unevaluated(self.tcx, ct);
-        let pattern = self.const_to_pat(c, ty, id, span);
-
-        // Apply a type ascription for the inline constant.
-        let annotation = {
-            let infcx = tcx.infer_ctxt().build(TypingMode::non_body_analysis());
-            let args = ty::InlineConstArgs::new(
-                tcx,
-                ty::InlineConstArgsParts { parent_args, ty: infcx.next_ty_var(span) },
-            )
-            .args;
-            infcx.canonicalize_user_type_annotation(ty::UserType::new(ty::UserTypeKind::TypeOf(
-                def_id.to_def_id(),
-                ty::UserArgs { args, user_self_ty: None },
-            )))
-        };
-        let annotation =
-            CanonicalUserTypeAnnotation { user_ty: Box::new(annotation), span, inferred_ty: ty };
-        PatKind::AscribeUserType {
-            subpattern: pattern,
-            ascription: Ascription {
-                annotation,
-                // Note that we use `Contravariant` here. See the `variance` field documentation
-                // for details.
-                variance: ty::Contravariant,
-            },
-        }
-    }
-
     /// Lowers the kinds of "expression" that can appear in a HIR pattern:
     /// - Paths (e.g. `FOO`, `foo::BAR`, `Option::None`)
-    /// - Inline const blocks (e.g. `const { 1 + 1 }`)
     /// - Literals, possibly negated (e.g. `-128u8`, `"hello"`)
     fn lower_pat_expr(
         &mut self,
@@ -677,9 +630,6 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
     ) -> PatKind<'tcx> {
         match &expr.kind {
             hir::PatExprKind::Path(qpath) => self.lower_path(qpath, expr.hir_id, expr.span).kind,
-            hir::PatExprKind::ConstBlock(anon_const) => {
-                self.lower_inline_const(anon_const, expr.hir_id, expr.span)
-            }
             hir::PatExprKind::Lit { lit, negated } => {
                 // We handle byte string literal patterns by using the pattern's type instead of the
                 // literal's type in `const_to_pat`: if the literal `b"..."` matches on a slice reference,

--- a/compiler/rustc_parse/src/parser/expr.rs
+++ b/compiler/rustc_parse/src/parser/expr.rs
@@ -1514,7 +1514,7 @@ impl<'a> Parser<'a> {
                     },
                 )
             } else if this.check_inline_const(0) {
-                this.parse_const_block(lo, false)
+                this.parse_const_block(lo)
             } else if this.may_recover() && this.is_do_catch_block() {
                 this.recover_do_catch()
             } else if this.is_try_block() {

--- a/compiler/rustc_parse/src/parser/mod.rs
+++ b/compiler/rustc_parse/src/parser/mod.rs
@@ -1300,7 +1300,7 @@ impl<'a> Parser<'a> {
     }
 
     /// Parses inline const expressions.
-    fn parse_const_block(&mut self, span: Span, pat: bool) -> PResult<'a, Box<Expr>> {
+    fn parse_const_block(&mut self, span: Span) -> PResult<'a, Box<Expr>> {
         self.expect_keyword(exp!(Const))?;
         let (attrs, blk) = self.parse_inner_attrs_and_block(None)?;
         let anon_const = AnonConst {
@@ -1308,18 +1308,7 @@ impl<'a> Parser<'a> {
             value: self.mk_expr(blk.span, ExprKind::Block(blk, None)),
         };
         let blk_span = anon_const.value.span;
-        let kind = if pat {
-            let guar = self
-                .dcx()
-                .struct_span_err(blk_span, "const blocks cannot be used as patterns")
-                .with_help(
-                    "use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead",
-                )
-                .emit();
-            ExprKind::Err(guar)
-        } else {
-            ExprKind::ConstBlock(anon_const)
-        };
+        let kind = ExprKind::ConstBlock(anon_const);
         Ok(self.mk_expr_with_attrs(span.to(blk_span), kind, attrs))
     }
 

--- a/compiler/rustc_parse/src/parser/pat.rs
+++ b/compiler/rustc_parse/src/parser/pat.rs
@@ -785,8 +785,10 @@ impl<'a> Parser<'a> {
         } else if self.eat_keyword(exp!(Box)) {
             self.parse_pat_box()?
         } else if self.check_inline_const(0) {
-            // Parse `const pat`
-            let const_expr = self.parse_const_block(lo.to(self.token.span), true)?;
+            // Parse `const pat`.
+            // NOTE: This will always error later during AST lowering because
+            // inline const cannot be used as patterns.
+            let const_expr = self.parse_const_block(lo.to(self.token.span))?;
 
             if let Some(re) = self.parse_range_end() {
                 self.parse_pat_range_begin_with(const_expr, re)?
@@ -1281,7 +1283,7 @@ impl<'a> Parser<'a> {
             .then_some(self.prev_token.span);
 
         let bound = if self.check_inline_const(0) {
-            self.parse_const_block(self.token.span, true)
+            self.parse_const_block(self.token.span)
         } else if self.check_path() {
             let lo = self.token.span;
             let (qself, path) = if self.eat_lt() {

--- a/src/tools/clippy/clippy_lints/src/utils/author.rs
+++ b/src/tools/clippy/clippy_lints/src/utils/author.rs
@@ -723,7 +723,6 @@ impl<'a, 'tcx> PrintVisitor<'a, 'tcx> {
                 kind!("Lit {{ ref {lit}, {negated} }}");
                 self.lit(lit);
             },
-            PatExprKind::ConstBlock(_) => kind!("ConstBlock(_)"),
             PatExprKind::Path(_) => self.maybe_path(pat),
         }
     }

--- a/src/tools/clippy/clippy_utils/src/consts.rs
+++ b/src/tools/clippy/clippy_utils/src/consts.rs
@@ -578,7 +578,6 @@ impl<'tcx> ConstEvalCtxt<'tcx> {
                     Some(val)
                 }
             },
-            PatExprKind::ConstBlock(ConstBlock { body, .. }) => self.expr(self.tcx.hir_body(*body).value),
             PatExprKind::Path(qpath) => self.qpath(qpath, pat_expr.hir_id),
         }
     }

--- a/src/tools/clippy/clippy_utils/src/hir_utils.rs
+++ b/src/tools/clippy/clippy_utils/src/hir_utils.rs
@@ -506,9 +506,8 @@ impl HirEqInterExpr<'_, '_, '_> {
                     negated: right_neg,
                 },
             ) => left_neg == right_neg && left.node == right.node,
-            (PatExprKind::ConstBlock(left), PatExprKind::ConstBlock(right)) => self.eq_body(left.body, right.body),
             (PatExprKind::Path(left), PatExprKind::Path(right)) => self.eq_qpath(left, right),
-            (PatExprKind::Lit { .. } | PatExprKind::ConstBlock(..) | PatExprKind::Path(..), _) => false,
+            (PatExprKind::Lit { .. } | PatExprKind::Path(..), _) => false,
         }
     }
 
@@ -1102,7 +1101,6 @@ impl<'a, 'tcx> SpanlessHash<'a, 'tcx> {
                 lit.node.hash(&mut self.s);
                 negated.hash(&mut self.s);
             },
-            PatExprKind::ConstBlock(c) => self.hash_body(c.body),
             PatExprKind::Path(qpath) => self.hash_qpath(qpath),
         }
     }

--- a/tests/ui/inline-const/in-pat-recovery.rs
+++ b/tests/ui/inline-const/in-pat-recovery.rs
@@ -4,8 +4,63 @@
 fn main() {
     match 1 {
         const { 1 + 7 } => {}
-        //~^ ERROR const blocks cannot be used as patterns
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
         2 => {}
+        _ => {}
+    }
+
+    match 5 {
+        const { 1 } ..= 10 => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        1 ..= const { 10 } => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        const { 1 } ..= const { 10 } => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        //~| ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        const { 1 } .. 10 => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        1 .. const { 10 } => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        const { 1 + 2 } ..= 10 => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        1 ..= const { 5 + 5 } => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        const { 3 } .. => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
+        _ => {}
+    }
+
+    match 5 {
+        ..= const { 7 } => {}
+        //~^ ERROR arbitrary expressions aren't allowed in patterns
         _ => {}
     }
 }

--- a/tests/ui/inline-const/in-pat-recovery.stderr
+++ b/tests/ui/inline-const/in-pat-recovery.stderr
@@ -1,10 +1,90 @@
-error: const blocks cannot be used as patterns
-  --> $DIR/in-pat-recovery.rs:6:15
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:6:9
    |
 LL |         const { 1 + 7 } => {}
-   |               ^^^^^^^^^
+   |         ^^^^^^^^^^^^^^^
    |
    = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
 
-error: aborting due to 1 previous error
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:13:9
+   |
+LL |         const { 1 } ..= 10 => {}
+   |         ^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:19:15
+   |
+LL |         1 ..= const { 10 } => {}
+   |               ^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:25:9
+   |
+LL |         const { 1 } ..= const { 10 } => {}
+   |         ^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:25:25
+   |
+LL |         const { 1 } ..= const { 10 } => {}
+   |                         ^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:32:9
+   |
+LL |         const { 1 } .. 10 => {}
+   |         ^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:38:14
+   |
+LL |         1 .. const { 10 } => {}
+   |              ^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:44:9
+   |
+LL |         const { 1 + 2 } ..= 10 => {}
+   |         ^^^^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:50:15
+   |
+LL |         1 ..= const { 5 + 5 } => {}
+   |               ^^^^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:56:9
+   |
+LL |         const { 3 } .. => {}
+   |         ^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/in-pat-recovery.rs:62:13
+   |
+LL |         ..= const { 7 } => {}
+   |             ^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error: aborting due to 11 previous errors
 

--- a/tests/ui/pattern/deref-patterns/ice-adjust-mode-unimplemented-for-constblock.rs
+++ b/tests/ui/pattern/deref-patterns/ice-adjust-mode-unimplemented-for-constblock.rs
@@ -1,0 +1,10 @@
+#![feature(deref_patterns)]
+#![expect(incomplete_features)]
+
+fn main() {
+    let vec![const { vec![] }]: Vec<usize> = vec![];
+    //~^ ERROR expected a pattern, found a function call
+    //~| ERROR usage of qualified paths in this context is experimental
+    //~| ERROR expected tuple struct or tuple variant
+    //~| ERROR arbitrary expressions aren't allowed in patterns
+}

--- a/tests/ui/pattern/deref-patterns/ice-adjust-mode-unimplemented-for-constblock.stderr
+++ b/tests/ui/pattern/deref-patterns/ice-adjust-mode-unimplemented-for-constblock.stderr
@@ -1,0 +1,41 @@
+error[E0532]: expected a pattern, found a function call
+  --> $DIR/ice-adjust-mode-unimplemented-for-constblock.rs:5:9
+   |
+LL |     let vec![const { vec![] }]: Vec<usize> = vec![];
+   |         ^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
+   |
+   = note: function calls are not allowed in patterns: <https://doc.rust-lang.org/book/ch19-00-patterns.html>
+   = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0658]: usage of qualified paths in this context is experimental
+  --> $DIR/ice-adjust-mode-unimplemented-for-constblock.rs:5:9
+   |
+LL |     let vec![const { vec![] }]: Vec<usize> = vec![];
+   |         ^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #86935 <https://github.com/rust-lang/rust/issues/86935> for more information
+   = help: add `#![feature(more_qualified_paths)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+   = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: arbitrary expressions aren't allowed in patterns
+  --> $DIR/ice-adjust-mode-unimplemented-for-constblock.rs:5:14
+   |
+LL |     let vec![const { vec![] }]: Vec<usize> = vec![];
+   |              ^^^^^^^^^^^^^^^^
+   |
+   = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
+
+error[E0164]: expected tuple struct or tuple variant, found associated function `<[_]>::into_vec`
+  --> $DIR/ice-adjust-mode-unimplemented-for-constblock.rs:5:9
+   |
+LL |     let vec![const { vec![] }]: Vec<usize> = vec![];
+   |         ^^^^^^^^^^^^^^^^^^^^^^ `fn` calls are not allowed in patterns
+   |
+   = help: for more information, visit https://doc.rust-lang.org/book/ch19-00-patterns.html
+   = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 4 previous errors
+
+Some errors have detailed explanations: E0164, E0532, E0658.
+For more information about an error, try `rustc --explain E0164`.

